src/ParseFasm.cpp

```diff
@@ -0,0 +1,508 @@
+        }
+
+        /// A Verilog hex value.
+        /// Tag: hex (h)
+        virtual Any visitHexValue(
+            FasmParser::HexValueContext* context) override {
+                std::ostringstream data;
+                std::string value = context->HEXADECIMAL_VALUE()->getText();
+                auto it = value.begin();
+                it += 2;  /// skip 'h
+
+                /// Build up Nums 4 bits at a time, skipping '_'.
+                int bits = lead_bits(count_without(it, value.end(), '_') * 4);
+                uint32_t word = 0;
+                while (it != value.end()) {
+                        if (*it != '_') {
+                                word = (word << 4) | from_hex_digit(*it);
+                                bits += 4;
+                                if (bits == Num::kWidth) {
+                                        data << Num(word);
+                                        word = 0;
+                                        bits = 0;
+                                }
+                        }
+                        it++;
+                }
+                assert(!word);
+                return withHeader(TAG('h', hex), data.str());
+        }
+
+        /// A Verilog binary value.
+        /// Tag: binary (b)
+        virtual Any visitBinaryValue(
+            FasmParser::BinaryValueContext* context) override {
+                std::ostringstream data;
+                std::string value = context->BINARY_VALUE()->getText();
+                auto it = value.begin();
+                it += 2;  /// skip 'b
+
+                /// Build up Nums a bit at a time, skipping '_'.
+                int bits = lead_bits(count_without(it, value.end(), '_'));
+                uint32_t word = 0;
+                while (it != value.end()) {
+                        if (*it != '_') {
+                                word = (word << 1) | (*it == '1' ? 1 : 0);
+                                bits += 1;
+                                if (bits == Num::kWidth) {
+                                        data << Num(word);
+                                        word = 0;
+                                        bits = 0;
+                                }
+                        }
+                        it++;
+                }
+                assert(!word);
+                return withHeader(TAG('b', binary), data.str());
+        }
+
+        /// A Verilog decimal value.
+        /// Tags: decimal (d)
+        virtual Any visitDecimalValue(
+            FasmParser::DecimalValueContext* context) override {
+                long long integer = 0;
+                std::string value = context->DECIMAL_VALUE()->getText();
+                auto it = value.begin();
+                it += 2;  /// skip 'd
+
+                /// Build up a Num, skipping '_'.
+                while (it != value.end()) {
+                        if (*it != '_') {
+                                integer = (integer * 10) + (*it - '0');
```

Comment:
This breaks on decimal values larger than sizeof(long long), yes?

Reply:
I've added an assertion to detect overflow. I haven't seen much use of large decimal numbers, and the implementation is somewhat tricky.